src/storage/index/mod.rs

```diff
@@ -0,0 +1,529 @@
+pub trait Key {
+    fn at(&self, pos: usize) -> u8;
+    fn len(&self) -> usize;
+    fn prefix_before(&self, length: usize) -> Self;
+    fn prefix_after(&self, start: usize) -> Self;
+    fn longest_common_prefix(&self, slice: &[u8]) -> usize;
+    fn as_slice(&self) -> &[u8];
+}
+
+pub trait KeyTrait:
+    Key + Clone + PartialEq + PartialOrd + Ord + Debug + for<'a> From<&'a [u8]>
+{
+}
+impl<T: Key + Clone + PartialOrd + PartialEq + Ord + Debug + for<'a> From<&'a [u8]>> KeyTrait
+    for T
+{
+}
+
+/*
+    Key trait implementations
+*/
+
+// Source: https://www.the-paper-trail.org/post/art-paper-notes/
+//
+// Keys can be of two types:
+// 1. Fixed-length datatypes such as 128-bit integers, or strings of exactly 64-bytes,
+// don’t have any problem because there can, by construction, never be any key that’s
+// a prefix of any other.
+//
+// 2. Variable-length datatypes such as general strings, can be transformed into types
+// where no key is the prefix of any other by a simple trick: append the NULL byte to every key.
+// The NULL byte, as it does in C-style strings, indicates that this is the end of the key, and
+// no characters can come after it. Therefore no string with a null-byte can be a prefix of any other,
+// because no string can have any characters after the NULL byte!
+//
+#[derive(Clone, Debug, Eq)]
+pub struct FixedKey<const SIZE: usize> {
+    content: [u8; SIZE],
+    len: usize,
+}
+
+impl<const SIZE: usize> PartialEq for FixedKey<SIZE> {
+    fn eq(&self, other: &Self) -> bool {
+        self.content[..self.len] == other.content[..other.len]
+    }
+}
+
+impl<const SIZE: usize> PartialOrd for FixedKey<SIZE> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl<const SIZE: usize> Ord for FixedKey<SIZE> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.content[..self.len].cmp(&other.content[..other.len])
+    }
+}
+
+impl<const SIZE: usize> FixedKey<SIZE> {
+    // Create new instance with data ending in zero byte
+    pub fn create_key(src: &[u8]) -> Self {
+        assert!(src.len() < SIZE);
+        let mut content = [0; SIZE];
+        content[..src.len()].copy_from_slice(src);
+        content[src.len()] = 0;
+        Self {
+            content,
+            len: src.len() + 1,
+        }
+    }
+
+    // Create new instance from slice
+    pub fn from_slice(src: &[u8]) -> Self {
+        assert!(src.len() <= SIZE);
+        let mut content = [0; SIZE];
+        content[..src.len()].copy_from_slice(src);
+        Self {
+            content,
+            len: src.len(),
+        }
+    }
+
+    pub fn from_str(s: &str) -> Self {
+        assert!(s.len() < SIZE, "data length is greater than array length");
+        let mut arr = [0; SIZE];
+        arr[..s.len()].copy_from_slice(s.as_bytes());
+        Self {
+            content: arr,
+            len: s.len() + 1,
+        }
+    }
```

Comment:
This can be confused with the standard trait `std::str::FromStr::from_str`. You may either rename the method, or implement the trait.

Reply:
noted, thank you for the information!